File: wds/chart_sonolus.py
# convert from chart_sonolus.c by LittleYang0531
# converted by t-wy
# used to host a sonolus server for testing official charts using the referenced engine
# Code Source:
# https://github.com/SonolusHaniwa/sonolus-sirius-engine/blob/master/convert.h#L120
# up to commit 11be3447bb33e1ebe7f3e9d4fa2b11c75cffeefd

class Note:
    def __init__(self):
        self.startTime = 0
        self.endTime = 0
        self.type = 0
        self.leftLane = 0
        self.laneLength = 0
        self.gimmickType = 0
        self.scratchLength = 0

    def __lt__(self, other):
        if self.endTime == other.endTime:
            return self.leftLane * 100 + self.laneLength < other.leftLane * 100 + other.laneLength
        return self.endTime < other.endTime

from enum import Enum, auto

class NoteType(Enum):
    HiSpeed = -1
    _None = 0
    Normal = 10
    Critical = 20
    Sound = 30
    SoundPurple = auto() # the original source swapped these two
    ScratchSound = 40
    Flick = 50
    HoldStart = 80
    CriticalHoldStart = auto()
    ScratchHoldStart = auto()
    ScratchCriticalHoldStart = auto()
    Hold = 100
    CriticalHold = auto()
    ScratchHold = 110
    ScratchCriticalHold = auto()
    HoldEighth = 900

def fromSirius(text: str, chartOffset: float, bgmOffset: float = 0) -> str:
    def add_log(*args):
        logging.info(" ".join(map(str, args)))
    from typing import Dict, List
    import json, logging
    # 谱面读取
    lines = text.split("\n")
    notes = []
    speed = 1

    for line in lines:
        x = Note()
        components = line.split(",")

        if len(components) < 7:
            break

        x.startTime = float(components[0]) + chartOffset
        x.endTime = float(components[1])
        if x.endTime != -1:
            x.endTime += chartOffset

        x.type = NoteType(int(components[2]))
        x.leftLane = int(components[3])
        x.laneLength = int(components[4])

        s = components[5]
        x.scratchLength = int(components[6])

        if s == "JumpScratch":
            x.gimmickType = 1 # "JumpScratch"
        elif s == "OneDirection":
            x.gimmickType = 2 # "OneDirection"
        elif all(c in "0123456789" for c in s):
            x.gimmickType = int(s)
        else:
            x.gimmickType = 0

        notes.append(x)

    notes.sort(key=lambda note: (note.startTime, note.type.value))
    add_log("[INFO] Total Note Number:", len(notes))

    # 开始转换
    res = []
    single = {}
    holdEnd: List[Note] = []
    SyncLineLeft: Dict[float, int] = {}
    SyncLineRight: Dict[float, int] = {}

    def addSyncLine(beat, leftLane, laneLength):
        if beat not in SyncLineLeft:
            SyncLineLeft[beat] = leftLane
        else:
            SyncLineLeft[beat] = min(SyncLineLeft[beat], leftLane)

        if beat not in SyncLineRight:
            SyncLineRight[beat] = leftLane + laneLength - 1
        else:
            SyncLineRight[beat] = max(SyncLineRight[beat], leftLane + laneLength - 1)

    res.append({
        "archetype": "Sirius Initialization",
        "data": []
    })
    res.append({
        "archetype": "Sirius Input Manager",
        "data": []
    })
    res.append({
        "archetype": "Sirius Stage",
        "data": []
    })
    res.append({
        "archetype": "#BPM_CHANGE",
        "data": [
            {"name": "#BEAT", "value": 0},
            {"name": "#BPM", "value": 60}
        ]
    })

    lastEighthTime = [[0 for _ in range(13)] for _ in range(13)]
    total = 0

    for i, note in enumerate(notes):
        # 提前处理 Sirius HoldEnd;
        while len(holdEnd) and holdEnd[0].endTime <= note.startTime:
            x = holdEnd.pop(0)
            single = {"data": []}
            if x.type == NoteType.Hold or x.type == NoteType.CriticalHold:
                single["archetype"] = "Sirius Hold End"
                single["data"].append({"name": "beat", "value": x.endTime})
                single["data"].append({"name": "stBeat", "value": x.startTime})
                single["data"].append({"name": "lane", "value": x.leftLane})
                single["data"].append({"name": "laneLength", "value": x.laneLength})
                total += 1
            else:
                single["archetype"] = "Sirius Scratch Hold End"
                single["data"].append({"name": "beat", "value": x.endTime})
                single["data"].append({"name": "stBeat", "value": x.startTime})
                single["data"].append({"name": "lane", "value": x.leftLane})
                single["data"].append({"name": "laneLength", "value": x.laneLength})
                single["data"].append({"name": "scratchLength", "value": x.scratchLength})
                total += 1
            res.append(single.copy())
            addSyncLine(x.endTime, x.leftLane, x.laneLength)
        # 处理当前 Note
        x = note
        single = {"data": []}
        if x.type == NoteType.Normal:
            single["archetype"] = "Sirius Normal Note"
            single["data"].append({"name": "beat", "value": x.startTime})
            single["data"].append({"name": "lane", "value": x.leftLane})
            single["data"].append({"name": "laneLength", "value": x.laneLength})
            addSyncLine(x.startTime, x.leftLane, x.laneLength)
            total += 1
        elif x.type == NoteType.Critical:
            single["archetype"] = "Sirius Critical Note"
            single["data"].append({"name": "beat", "value": x.startTime})
            single["data"].append({"name": "lane", "value": x.leftLane})
            single["data"].append({"name": "laneLength", "value": x.laneLength})
            addSyncLine(x.startTime, x.leftLane, x.laneLength)
            total += 1
        elif x.type == NoteType.Flick:
            single["archetype"] = "Sirius Flick Note"
            single["data"].append({"name": "beat", "value": x.startTime})
            single["data"].append({"name": "lane", "value": x.leftLane})
            single["data"].append({"name": "laneLength", "value": x.laneLength})
            single["data"].append({"name": "scratchLength", "value": x.scratchLength})
            addSyncLine(x.startTime, x.leftLane, x.laneLength)
            total += 1
        elif x.type == NoteType.HoldStart:
            single["archetype"] = "Sirius Hold Start"
            single["data"].append({"name": "beat", "value": x.startTime})
            single["data"].append({"name": "lane", "value": x.leftLane})
            single["data"].append({"name": "laneLength", "value": x.laneLength})
            addSyncLine(x.startTime, x.leftLane, x.laneLength)
            total += 1
        elif x.type == NoteType.CriticalHoldStart:
            single["archetype"] = "Sirius Critical Hold Start"
            single["data"].append({"name": "beat", "value": x.startTime})
            single["data"].append({"name": "lane", "value": x.leftLane})
            single["data"].append({"name": "laneLength", "value": x.laneLength})
            addSyncLine(x.startTime, x.leftLane, x.laneLength)
            total += 1
        elif x.type == NoteType.ScratchHoldStart:
            single["archetype"] = "Sirius Scratch Hold Start"
            single["data"].append({"name": "beat", "value": x.startTime})
            single["data"].append({"name": "lane", "value": x.leftLane})
            single["data"].append({"name": "laneLength", "value": x.laneLength})
            addSyncLine(x.startTime, x.leftLane, x.laneLength)
            total += 1
        elif x.type == NoteType.ScratchCriticalHoldStart:
            single["archetype"] = "Sirius Critical Scratch Hold Start"
            single["data"].append({"name": "beat", "value": x.startTime})
            single["data"].append({"name": "lane", "value": x.leftLane})
            single["data"].append({"name": "laneLength", "value": x.laneLength})
            addSyncLine(x.startTime, x.leftLane, x.laneLength)
            total += 1
        elif x.type in [NoteType.Hold, NoteType.CriticalHold, NoteType.ScratchHold, NoteType.ScratchCriticalHold]:
            holdEnd.append(x)
        elif x.type in [NoteType.Sound, NoteType.SoundPurple]:
            lastEighthTime[x.leftLane][x.leftLane + x.laneLength - 1] = x.startTime
            single["archetype"] = "Sirius Sound"
            single["data"].append({"name": "beat", "value": x.startTime})
            single["data"].append({"name": "lane", "value": x.leftLane})
            single["data"].append({"name": "laneLength", "value": x.laneLength})
            single["data"].append({"name": "holdType", "value": NoteType.Hold.value if x.type == NoteType.Sound else NoteType.ScratchHold.value})
            total += 1
        elif x.type == NoteType.ScratchSound:
            lastEighthTime[x.leftLane][x.leftLane + x.laneLength - 1] = x.startTime
            single["archetype"] = "Sirius Scratch Hold End"
            single["data"].append({"name": "beat", "value": x.startTime})
            single["data"].append({"name": "stBeat", "value": x.startTime})
            single["data"].append({"name": "lane", "value": x.leftLane})
            single["data"].append({"name": "laneLength", "value": x.laneLength})
            single["data"].append({"name": "scratchLength", "value": 0})
            addSyncLine(x.startTime, x.leftLane, x.laneLength)
            total += 1
        elif x.type == NoteType.HoldEighth:
            if lastEighthTime[x.leftLane][x.leftLane + x.laneLength - 1] == x.startTime:
                continue
            single["archetype"] = "Sirius Hold Eighth"
            single["data"].append({"name": "beat", "value": x.startTime})
            single["data"].append({"name": "lane", "value": x.leftLane})
            single["data"].append({"name": "laneLength", "value": x.laneLength})
            single["data"].append({"name": "holdType", "value": 0})
            total += 1
        elif x.type == NoteType._None:
            single["archetype"] = "Sirius Split Line"
            single["data"].append({"name": "beat", "value": x.startTime})
            single["data"].append({"name": "endBeat", "value": x.endTime})
            single["data"].append({"name": "split", "value": x.gimmickType % 10})
            single["data"].append({"name": "color", "value": x.scratchLength})
        elif x.type == NoteType.HiSpeed:
            single["archetype"] = "#TIMESCALE_CHANGE"
            single["data"].append({"name": "#BEAT", "value": x.startTime})
            single["data"].append({"name": "#TIMESCALE", "value": x.endTime - chartOffset})

        if "archetype" in single:
            res.append(single)
        else:
            add_log(single)

        if int(10 * (i + 1) / len(notes)) != int(10 * i / len(notes)):
            add_log("[INFO]", 100 * (i + 1) / len(notes), "% Notes Solved.")

    while holdEnd:
        x = holdEnd.pop(0)
        single = {"data": []}
        if x.type == NoteType.Hold or x.type == NoteType.CriticalHold:
            single["archetype"] = "Sirius Hold End"
            single["data"].append({"name": "beat", "value": x.endTime})
            single["data"].append({"name": "stBeat", "value": x.startTime})
            single["data"].append({"name": "lane", "value": x.leftLane})
            single["data"].append({"name": "laneLength", "value": x.laneLength})
            total += 1
        else:
            single["archetype"] = "Sirius Scratch Hold End"
            single["data"].append({"name": "beat", "value": x.endTime})
            single["data"].append({"name": "stBeat", "value": x.startTime})
            single["data"].append({"name": "lane", "value": x.leftLane})
            single["data"].append({"name": "laneLength", "value": x.laneLength})
            single["data"].append({"name": "scratchLength", "value": x.scratchLength})
            total += 1

        res.append(single)
        addSyncLine(x.endTime, x.leftLane, x.laneLength)

    add_log("[INFO] Total Note Number:", total)
    add_log("[INFO] Solving Sync Line...")

    for beat, left in SyncLineLeft.items():
        single = {"data": []}
        single["archetype"] = "Sirius Sync Line"
        single["data"].append({"name": "beat", "value": beat})
        single["data"].append({"name": "left", "value": left})
        single["data"].append({"name": "right", "value": SyncLineRight[beat]})
        res.append(single.copy())

    add_log("[INFO] Sync Line Solved.")

    data = {
        "formatVersion": 5,
        "bgmOffset": bgmOffset,
        "entities": res
    }

    return json.dumps(data, separators=(',', ':'))

if __name__ == "__main__":
    import logging
    logging.basicConfig(level=logging.DEBUG)
    text = """
<insert test chart>
""".strip()
    chartOffset = 0.5
    bgmOffset = 0
    result = fromSirius(text, chartOffset, bgmOffset)
    # print(result)

File: wds/chart_sonolus.c
// https://github.com/SonolusHaniwa/sonolus-sirius-engine/blob/master/convert.h#L120
// up to commit 11be3447bb33e1ebe7f3e9d4fa2b11c75cffeefd

string fromSirius(string text, double chartOffset, double bgmOffset = 0) {
    // 谱面读取
    auto lines = explode("\n", text.c_str());
    vector<Note> notes; double speed = 1;
    for (int i = 0; i < lines.size(); i++) {
        char unused = 0;
        Note x; string xx = lines[i];
        auto components = explode(",", xx.c_str());
        if (components.size() < 7) break;
        x.startTime = atof(components[0].c_str()) + chartOffset;
        x.endTime = atof(components[1].c_str());
        if (x.endTime != -1) x.endTime += chartOffset;
        x.type = atoi(components[2].c_str());
        x.leftLane = atoi(components[3].c_str());
        x.laneLength = atoi(components[4].c_str());
        string s = components[5];
        x.scratchLength = atoi(components[6].c_str());
        if (s == "JumpScratch") x.gimmickType = JumpScratch;
        else if (s == "OneDirection") x.gimmickType = OneDirection;
        else if (s == "11") x.gimmickType = Split1;
        else if (s == "12") x.gimmickType = Split2;
        else if (s == "13") x.gimmickType = Split3;
        else if (s == "14") x.gimmickType = Split4;
        else if (s == "15") x.gimmickType = Split5;
        else if (s == "16") x.gimmickType = Split6;
        else x.gimmickType = 0;
        notes.push_back(x);
    } sort(notes.begin(), notes.end(), [&](Note a, Note b){
        if (a.startTime == b.startTime) return a.type < b.type;
        return a.startTime < b.startTime;
    });
    cout << "[INFO] Total Note Number: " << notes.size() << endl;

    // 开始转换
	Json::Value res, single; set<Note> holdEnd;
    map<double, int> SyncLineLeft, SyncLineRight;
    auto addSyncLine = [&](double beat, int leftLane, int laneLength) {
        if (SyncLineLeft.find(beat) == SyncLineLeft.end()) SyncLineLeft[beat] = leftLane;
        else SyncLineLeft[beat] = min(SyncLineLeft[beat], leftLane);
        if (SyncLineRight.find(beat) == SyncLineRight.end()) SyncLineRight[beat] = leftLane + laneLength - 1;
        else SyncLineRight[beat] = max(SyncLineRight[beat], leftLane + laneLength - 1);
    };
	single["archetype"] = "Sirius Initialization"; single["data"].resize(0);
	res.append(single);
	single["archetype"] = "Sirius Input Manager"; single["data"].resize(0);
	res.append(single);
	single["archetype"] = "Sirius Stage"; single["data"].resize(0);
	res.append(single); 
	single["archetype"] = "#BPM_CHANGE";
	single["data"][0]["name"] = "#BEAT"; single["data"][0]["value"] = 0;
	single["data"][1]["name"] = "#BPM"; single["data"][1]["value"] = 60;
	res.append(single); 
    double lastEighthTime[13][13]; int total = 0;
    for (int i = 0; i < 13; i++) for (int j = 0; j < 13; j++) lastEighthTime[i][j] = 0;
    for (int i = 0; i < notes.size(); i++) {
        // 提前处理 Sirius HoldEnd;
        while (holdEnd.size() && (*holdEnd.begin()).endTime <= notes[i].startTime) {
            Note x = *holdEnd.begin(); holdEnd.erase(holdEnd.begin()); Json::Value single;
            if (x.type == Hold || x.type == CriticalHold) {
                single["archetype"] = "Sirius Hold End";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.endTime;
                single["data"][1]["name"] = "stBeat"; single["data"][1]["value"] = x.startTime;
                single["data"][2]["name"] = "lane"; single["data"][2]["value"] = x.leftLane;
                single["data"][3]["name"] = "laneLength"; single["data"][3]["value"] = x.laneLength;
                total++;
            } else {
                single["archetype"] = "Sirius Scratch Hold End";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.endTime;
                single["data"][1]["name"] = "stBeat"; single["data"][1]["value"] = x.startTime;
                single["data"][2]["name"] = "lane"; single["data"][2]["value"] = x.leftLane;
                single["data"][3]["name"] = "laneLength"; single["data"][3]["value"] = x.laneLength;
                single["data"][4]["name"] = "scratchLength"; single["data"][4]["value"] = x.scratchLength;
                total++;
            }
            res.append(single);
			addSyncLine(x.endTime, x.leftLane, x.laneLength);
        }
        // 处理当前 Note
        Note x = notes[i]; Json::Value single;
        single["archetype"] = "";
        switch(x.type) {
            case Normal: {
                single["archetype"] = "Sirius Normal Note";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.startTime;
                single["data"][1]["name"] = "lane"; single["data"][1]["value"] = x.leftLane;
                single["data"][2]["name"] = "laneLength"; single["data"][2]["value"] = x.laneLength;
                addSyncLine(x.startTime, x.leftLane, x.laneLength);
                total++;
            } break;
            case Critical: {
                single["archetype"] = "Sirius Critical Note";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.startTime;
                single["data"][1]["name"] = "lane"; single["data"][1]["value"] = x.leftLane;
                single["data"][2]["name"] = "laneLength"; single["data"][2]["value"] = x.laneLength;
                addSyncLine(x.startTime, x.leftLane, x.laneLength);
                total++;
            } break;
            case Flick: {
                single["archetype"] = "Sirius Flick Note";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.startTime;
                single["data"][1]["name"] = "lane"; single["data"][1]["value"] = x.leftLane;
                single["data"][2]["name"] = "laneLength"; single["data"][2]["value"] = x.laneLength;
                single["data"][3]["name"] = "scratchLength"; single["data"][3]["value"] = x.scratchLength;
                addSyncLine(x.startTime, x.leftLane, x.laneLength);
                total++;
            } break;
            case HoldStart: {
                single["archetype"] = "Sirius Hold Start";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.startTime;
                single["data"][1]["name"] = "lane"; single["data"][1]["value"] = x.leftLane;
                single["data"][2]["name"] = "laneLength"; single["data"][2]["value"] = x.laneLength;
                addSyncLine(x.startTime, x.leftLane, x.laneLength);
                total++;
            } break; 
            case CriticalHoldStart: {
                single["archetype"] = "Sirius Critical Hold Start";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.startTime;
                single["data"][1]["name"] = "lane"; single["data"][1]["value"] = x.leftLane;
                single["data"][2]["name"] = "laneLength"; single["data"][2]["value"] = x.laneLength;
                addSyncLine(x.startTime, x.leftLane, x.laneLength);
                total++;
            } break;
            case ScratchHoldStart: {
                single["archetype"] = "Sirius Scratch Hold Start";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.startTime;
                single["data"][1]["name"] = "lane"; single["data"][1]["value"] = x.leftLane;
                single["data"][2]["name"] = "laneLength"; single["data"][2]["value"] = x.laneLength;
                addSyncLine(x.startTime, x.leftLane, x.laneLength);
                total++;
            } break;
            case ScratchCriticalHoldStart: {
                single["archetype"] = "Sirius Critical Scratch Hold Start";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.startTime;
                single["data"][1]["name"] = "lane"; single["data"][1]["value"] = x.leftLane;
                single["data"][2]["name"] = "laneLength"; single["data"][2]["value"] = x.laneLength;
                addSyncLine(x.startTime, x.leftLane, x.laneLength);
                total++;
            } break;
            case Hold: case CriticalHold: case ScratchHold: case ScratchCriticalHold: {
                holdEnd.insert(x);
            } break;
            case Sound: case ScratchSound: {
                lastEighthTime[x.leftLane][x.leftLane + x.laneLength - 1] = x.startTime;
                single["archetype"] = "Sirius Sound";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.startTime;
                single["data"][1]["name"] = "lane"; single["data"][1]["value"] = x.leftLane;
                single["data"][2]["name"] = "laneLength"; single["data"][2]["value"] = x.laneLength;
				single["data"][3]["name"] = "holdType"; single["data"][3]["value"] = Sound ? Hold : ScratchHold;
                total++;
            } break;
            case SoundPurple: {
                lastEighthTime[x.leftLane][x.leftLane + x.laneLength - 1] = x.startTime;
                single["archetype"] = "Sirius Scratch Hold End";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.startTime;
            	single["data"][1]["name"] = "stBeat"; single["data"][1]["value"] = x.startTime;
                single["data"][2]["name"] = "lane"; single["data"][2]["value"] = x.leftLane;
                single["data"][3]["name"] = "laneLength"; single["data"][3]["value"] = x.laneLength;
                single["data"][4]["name"] = "scratchLength"; single["data"][4]["value"] = 0;
                addSyncLine(x.startTime, x.leftLane, x.laneLength);
                total++;
            } break;
            case HoldEighth: {
                if (lastEighthTime[x.leftLane][x.leftLane + x.laneLength - 1] == x.startTime) break;
                single["archetype"] = "Sirius Hold Eighth";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.startTime;
                single["data"][1]["name"] = "lane"; single["data"][1]["value"] = x.leftLane;
                single["data"][2]["name"] = "laneLength"; single["data"][2]["value"] = x.laneLength;
                // 历史遗留代码了，这个属性现在应该不需要了 2024.5.27
				single["data"][3]["name"] = "holdType"; single["data"][3]["value"] = 0;
                total++;
            } break;
            case None: {
                single["archetype"] = "Sirius Split Line";
                single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.startTime;
                single["data"][1]["name"] = "endBeat"; single["data"][1]["value"] = x.endTime;
                single["data"][2]["name"] = "split"; single["data"][2]["value"] = x.gimmickType - 10;
                single["data"][3]["name"] = "color"; single["data"][3]["value"] = x.scratchLength;
            } break;
            case HiSpeed: {
                single["archetype"] = "#TIMESCALE_CHANGE";
                single["data"][0]["name"] = "#BEAT"; single["data"][0]["value"] = x.startTime;
                single["data"][1]["name"] = "#TIMESCALE"; single["data"][1]["value"] = x.endTime - chartOffset;
            }
        } if (single["archetype"] != "") res.append(single);
        if (single["archetype"] == Json::Value::null) cout << single << endl;
        if ((10 * (i + 1) / notes.size()) != (10 * i / notes.size())) cout << "[INFO] " << 100 * (i + 1) / notes.size() << "% Notes Solved." << endl;
    }

	// 处理完 HoldEnd
	while (holdEnd.size()) {
        Note x = *holdEnd.begin(); holdEnd.erase(holdEnd.begin()); Json::Value single;
        if (x.type == Hold || x.type == CriticalHold) {
            single["archetype"] = "Sirius Hold End";
            single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.endTime;
            single["data"][1]["name"] = "stBeat"; single["data"][1]["value"] = x.startTime;
            single["data"][2]["name"] = "lane"; single["data"][2]["value"] = x.leftLane;
            single["data"][3]["name"] = "laneLength"; single["data"][3]["value"] = x.laneLength;
            total++;
        } else {
            single["archetype"] = "Sirius Scratch Hold End";
            single["data"][0]["name"] = "beat"; single["data"][0]["value"] = x.endTime;
            single["data"][1]["name"] = "stBeat"; single["data"][1]["value"] = x.startTime;
            single["data"][2]["name"] = "lane"; single["data"][2]["value"] = x.leftLane;
            single["data"][3]["name"] = "laneLength"; single["data"][3]["value"] = x.laneLength;
            single["data"][4]["name"] = "scratchLength"; single["data"][4]["value"] = x.scratchLength;
            total++;
        } res.append(single);
        addSyncLine(x.endTime, x.leftLane, x.laneLength);
    }

    cout << "[INFO] Total Note Number: " << total << endl;
    cout << "[INFO] Solving Sync Line..." << endl;

    // 处理同步线
    for (auto v : SyncLineLeft) {
        double beat = v.first; int left = v.second;
        Json::Value single; single["archetype"] = "Sirius Sync Line";
        single["data"][0]["name"] = "beat"; single["data"][0]["value"] = beat;
        single["data"][1]["name"] = "left"; single["data"][1]["value"] = left;
        single["data"][2]["name"] = "right"; single["data"][2]["value"] = SyncLineRight[beat];
        res.append(single);
    }

    cout << "[INFO] Sync Line Solved." << endl;

	Json::Value data;
	data["formatVersion"] = 5;
	data["bgmOffset"] = bgmOffset;
	data["entities"] = res;
	return json_encode(data);
}